Handle invalid map data and out-of-bounds objects more gracefully

MichalZr6 opened the pull request:
Fixes: 
- #2166

<html><head></head><body><h3>Summary</h3><p>This PR hardens VCMI against invalid or extended H3M map data and avoids several crashes that occurred on community / HotA-style maps.</p><hr><h3>Key changes</h3><ul><li><p><strong>Map parsing robustness</strong></p><ul><li><p><code inline="">CBinaryReader::readBaseString()</code> now throws <code inline="">std::runtime_error()</code> when encountering an absurd string length instead of tripping a debug assert and crashing.</p></li><li><p>Callers can report a clean “Invalid or corrupt map” error to the user (similar to HotA) instead of failing with a crash</p></li></ul></li><li><p><strong>Safer object type changes</strong></p><ul><li><p><code inline="">CGObjectInstance::setType()</code>:</p><ul><li><p>Only calls <code inline="">getCornerOffset()</code> when an offset adjustment is actually needed (prisons → hero, monsters, creature generators), avoiding asserts on non-visitable templates (e.g. resource piles).</p></li><li><p>Skips appearance updates when <code inline="">cb-&gt;getTile(visitablePos())</code> returns <code inline="">nullptr</code>, logging a warning instead of dereferencing a null tile. Edit: should not happen now with the approach "push visitablePos into map bounds").</p></li></ul></li></ul></li><li><p><strong>Out-of-bounds visitable positions</strong></p><ul><li><p>Update: Objects that are visitable and placed outside of map's bounds are now <i>pushed</i> into map's boundaries before initialisation step.</p></li></ul></li><li><p><strong>Special loss conditions with placeholder heroes</strong></p><ul><li><p>Maps where “losing &lt;hero&gt;” is marked as a special loss condition but that hero is only used as a <strong>placeholder</strong> (uninitialized / unflagged owner) <s>will still cause instant loss on day 1 but not crash - those cases need to be handled still. HOTA allows such map to be played though.</s> are now handled and playable. The placeholder hero is discarded, in-line with OH3 behavior.</p></li></ul></li></ul><hr><h3>Testing</h3><p>Maps were tested one by one from the below list, top to bottom and some of them became playable due to fixes done along the way; they are still listed here for completeness.</p>

<h4>Map status after this PR</h4>

<table>
  <thead>
    <tr>
      <th>Map name</th>
      <th>Status / notes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>AB_Angor_s_Wrath.h3m</td>
      <td>Losing Sir Mullich is marked as a special loss condition but he is only a hero placeholder. OH3 accepts this; VCMI <s>triggers insta-loss.</s> now also allows this as well.</td>
    </tr>
    <tr>
      <td>AB_Attainment INCORRECT DATA CHECK.h3m</td>
      <td>Reported as “Invalid or corrupt map” instead of crashing. OH3 - same behavior in this case.</td>
    </tr>
    <tr>
      <td>AB_A_War INCORRECT DATA CHECK.h3m</td>
      <td>Playable (already fixed by an earlier PR).</td>
    </tr>
    <tr>
      <td>AB_Theophobia TAKE RESOURSES OR VISIT STAR AXES.h3m</td>
      <td>Playable with current <code>develop</code>.</td>
    </tr>
    <tr>
      <td>Baker Street (CHI).h3m</td>
      <td>Playable with current <code>develop</code>.</td>
    </tr>
    <tr>
      <td>Brave Heart (CHI).h3m</td>
      <td>Playable with current <code>develop</code>.</td>
    </tr>
    <tr>
      <td>Brave People-BT (CHI).h3m</td>
      <td>Playable with current <code>develop</code>.</td>
    </tr>
    <tr>
      <td>C-_Shadows_of_the_Forest ENG.h3m</td>
      <td>Previously crashed due to a resource pile placed at (108, 0) out of map bounds (invalid visitable pos). Now logs a warning and <s>skips that object.</s> moves that object visitablePos into the map bounds.</td>
    </tr>
    <tr>
      <td>Defend Taierzhuang (CHI).h3m</td>
      <td>Previously crashed due to multiple objects with out-of-bounds visitable positions. Now logs and <s>skips</s> moves those problematic objects; map is playable.</td>
    </tr>
    <tr>
      <td>Disappearing Planning Room (CHI).h3m</td>
      <td>Previously crashed; now playable.</td>
    </tr>
    <tr>
      <td>Doomsday Battle-BT (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Dragon's Nest (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Jedi Strikes Back (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Jedi Strikes Back-BT (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Jennifer's Dream Journey (CHI) ПОРАЖЕНИЕ LOSE.h3m</td>
      <td><s>Results in insta-loss.</s> Update: playable but with lots of objects not reachable for AI.</td>
    </tr>
    <tr>
      <td>Liu Bang’s Story (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Matrix (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Niklots Testament.h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>_Rise of the Sun King_ v26.0f (E) - UNABLE TO SELECT RANDOM ITEM FROM EMPTY CONTAINER.h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Saint Seiya the Goddess of Wisdom-BT (CHI).h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>SoD_Dart_NeoEdik__часть_2.h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
    <tr>
      <td>Where the beauty.h3m</td>
      <td>Playable (not tested in current develop)</td>
    </tr>
  </tbody>
</table>


<blockquote><p>Note: this PR primarily ensures the above community maps no longer crash due to invalid data and that VCMI reports invalid maps clearly instead of crashing.</p></blockquote></body></html>

MichalZr6 committed:
Changes making some community maps playable or at least preventing crashes running them

MichalZr6 committed:
Make MapReaderH3M::readBool tolerant to invalid values

MichalZr6 committed:
Avoid instant defeat on neutral/placeholder town & hero being set as loss cond.

IvanSavenko commented on dependencies:
Looks like unintended change. Please revert

MichalZr6 commented:
I rebased the branch onto `develop` and updated the PR target accordingly.

@IvanSavenko, could you please review when you have a moment?

Is anything else needed from my side here?

IvanSavenko commented on lib/mapping/MapReaderH3M.cpp:
```diff
@@ -479,10 +520,17 @@ void MapReaderH3M::readResources(TResources & resources)
-	assert(result == 0 || result == 1);
+	const uint8_t raw = readUInt8();
+
+	if(raw != 0 && raw != 1)
+	{
+		logGlobal->warn(
+			"MapReaderH3M: invalid bool value %u in H3M data, using LSB (%u) as value",
+			static_cast<unsigned>(raw),
+			static_cast<unsigned>(raw & 1));
+	}
```
Do we know in which cases this happens? Are you sure that this is not incorrect parsing of hota/wog map format?

MichalZr6 replied:
A map not from the test-pack, named `A Warm and Familiar Place` triggers the assert. Probably no impact on release builds as commenting out the assert allows to run this map without problems.

IvanSavenko replied:
Isn't that standard map? Or some modified version of it?

IvanSavenko commented on lib/mapping/MapReaderH3M.cpp:
```diff
@@ -200,24 +200,65 @@ FactionID MapReaderH3M::readFaction32()
+
+	if(type > features.riversCount)
+	{
+		// Map uses extended river type not supported by our config
```
Are these coming from wog map editor?

Also - do we know how h3 handles this?

MichalZr6 replied:
The assert was triggered specifically for `A War` map from the test-pack but on current develop debug run with this map is:
```cpp 
const uint8_t raw = readInt8();
uint8_t type = raw & 0x07;
raw == 0
type == 0
```
and that guard is unused.

IvanSavenko commented on lib/gameState/CGameState.cpp:
```diff
@@ -538,6 +578,25 @@ void CGameState::randomizeMapObjects(IGameRandomizer & gameRandomizer)
+		}
+	}
+}
+
```
Why not simply mark all objects owned by players on start?

MichalZr6 replied:
Done

IvanSavenko commented on lib/gameState/CGameState.h:
```diff
@@ -263,6 +272,8 @@ class DLL_LINKAGE CGameState : public CNonConstInfoCallback, public Serializeabl
 	Services * services;
 
+	std::array<std::set<ObjectInstanceID>, PlayerColor::PLAYER_LIMIT_I> everControlledObjects;
+
```
Perhaps should be part of player state?

MichalZr6 replied:
Done

IvanSavenko commented on lib/gameState/CGameState.cpp:
```diff
@@ -1235,14 +1344,28 @@ EVictoryLossCheckResult CGameState::checkForVictoryAndLoss(const PlayerColor & p
-			if (event.effect.type == EventEffect::DEFEAT)
+				// If our custom logic says "no loss yet", do NOT fall through to generic .test()
+				continue;
+			}
+
+			// All other defeat events: generic evaluation
+			if(event.trigger.test(evaluateEvent))
```
Perhaps we should split existing "control" condition into:
- control: returns true if controls or not has ever controlled
- capture: returns true if controls

For heroes - use existing 'control', for conditions like capture town - use 'capture'

Should simplify this logic a bit

MichalZr6 replied:
Good idea.
I used `CONTROL_CURRENT` rather than `CAPTURE`, because this is really a strict ownership check, not a capture event.

Current semantics are:
  - `CONTROL_CURRENT`: true only while the player/team currently controls the target
  - `CONTROL`: for a specific object, true until it was controlled and then lost; for generic type checks -> same behavior as `CONTROL_CURRENT`

Map objective resolution now rewrites victory-style `CONTROL` objectives to `CONTROL_CURRENT`, which matches "currently controls this town/object" more precisely.

IvanSavenko commented on lib/mapping/CMap.cpp:
```diff
@@ -414,10 +414,18 @@ const CGObjectInstance * CMap::getObjectiveObjectFrom(const int3 & pos, Obj type
+
+	// Hero special case: this often means a placeholder hero used.
+	// OH3 treats such conditions as inactive if the hero doesn't actually exist on the map,
+	// so don't try to substitute some random other hero as it will trigger insta-loss.
```
Perhaps it is not that special? And can use "has ever controlled" logic from this PR?

So, if placeholder for specific type was not replaced - condition will be inactive for entire map due to hero not present in 'has ever controlled' list?

MichalZr6 replied:
This is done, but not via `hasEverControlled` alone.

The important part is that hero placeholders are cleared right in first lines of `CGameState::init` after they are turned into real heroes for campaign cases.
So if we leave such an objective as a normal hero target there is no real object to bind it to later. I added `map->resolveHeroPlaceholderObjectives()` before placeholder removal.
So the flow is:

  - detect hero-placeholder-targeted objectives while placeholders still exist
  - rewrite/mark them explicitly
  - then remove placeholders as usual

After that, those conditions stay intentionally inactive instead of resolving to some unrelated hero or causing instant loss. `hasEverControlled` is part of the control semantics work, but the key fix for placeholder heroes was resolving them before they disappear.

IvanSavenko commented on server/processors/NewTurnProcessor.cpp:
```diff
@@ -90,7 +90,8 @@ void NewTurnProcessor::handleTownEvents(const CGTownInstance * town)
 		PlayerColor player = town->getOwner();
-		if (!event.affectsPlayer(player, gameHandler->gameInfo().getPlayerState(player)->isHuman()))
+		const auto * playerState = gameHandler->gameInfo().getPlayerState(player);
+		if(playerState && !event.affectsPlayer(player, playerState->isHuman()))
```
Should never happen I think? Town events are executed on start of corresponding player turn. As result - events in neutral towns never trigger

MichalZr6 replied:
I'm reverting this. I remember there was a case like this and the guard prevented the crash, but can't say which map caused that. Checked all of them from test-pack with legacy code and there was no crash.

IvanSavenko reviewed (changes requested):
Waiting for resolving of open review points + conflicts

MichalZr6 commented:
> Waiting for resolving of open review points + conflicts

Working on it

MichalZr6 committed:
Track object control history in PlayerState

MichalZr6 committed:
Separate current ownership objectives from historical control checks

MichalZr6 committed:
Fix whitespaces

MichalZr6 committed:
Make objective target lookup strict

MichalZr6 committed:
Handle hero-placeholder win/loss objectives

Maps that use hero placeholders in special victory/loss conditions are playable in HotA/OH3.
Handle those objectives explicitly in VCMI so such maps remain playable.

GrafDeVan commented:
Long time ago I updated my MapPackforDevelopers2. In current develop every map doesn't start or crashes. 
 
[MapPackforDevelopers 2.zip](https://github.com/user-attachments/files/28389210/MapPackforDevelopers.2.zip)

(sry, but I cant find your PR6438 in github actions)

MichalZr6 commented:
I will update this PR today. Got to push ready and working code for hero placeholders being set as map's win/loss objectives. For example `Angor's Wrath` map case.

GrafDeVan commented:
> I will update this PR today. Got to push ready and working code for hero placeholders being set as map's win/loss objectives. For example `Angor's Wrath` map case.

I've tested every map, remaine only who crashes
[MapPackforDevelopers 3.zip](https://github.com/user-attachments/files/28398518/MapPackforDevelopers.3.zip)

MichalZr6 commented:
Thanks a lot, much appreciated for your effort. Started checking those maps but had run out of time. Will check those two that left and address the review next.

GrafDeVan commented:
and 1 thing still here, about Angor's Wrath

here how game started
<img width="1920" height="1080" alt="изображение" src="https://github.com/user-attachments/assets/ed88377f-b95c-433b-b619-65e5c0bfe11c" />

and here how it in the map editor
<img width="262" height="260" alt="изображение" src="https://github.com/user-attachments/assets/9dc03339-1dea-40b1-bd02-1c16d27497da" />

MichalZr6 commented:
Same behavior as OH3. 
The only thing I added (and is different in that case) is a VCMI map's validator warning for such hero placeholder being set as victory / loss condition.

GrafDeVan commented:
> Same behavior as OH3. The only thing I added (and is different in that case) is a VCMI map's validator warning for such hero placeholder being set as victory / loss condition.

Okay. I don't even understand how this works in the original. The hero doesn't appear. What's the point then? Put him on the map. There's not even a tavern in the town.

misiokles commented:
I can see at screens, there's placeholder hero that can pass between scenarios.

MichalZr6 committed:
Improve diagnostic logs for invalid creature IDs during H3M map loading

GrafDeVan commented:
edited, my bad
@MichalZr6 Unfortunately, I have Windows 10 x64, and one map close the game 'SoD_Corbus_Realm_2'. And two don't start at all 'Rise of the Sun King_ v26.0f (E)', 'Attainment'. Rest doesn't work. dfebca668ef70756c189b801c61498d234b090f0,

MichalZr6 commented:
`Rise of the Sun King_ v26.0f (E)` is a map created with WOG mapeditor with a bunch of errors. I prepared and made a fixup for code in [a14a10f](https://github.com/vcmi/vcmi/pull/6438/commits/a14a10fa879956ba44a75504b7025c0a15816e5c) so it becomes playable.

`Attainment` is broken and unplayable in OH3 / HoTA. In VCMI it should now emit a window message for such case instead of crashing the client.

The third mentioned I haven't checked yet. Will do later.

> Rest doesn't work. 

Do you mean all other maps are also not playable ?

GrafDeVan commented:
> Do you mean all other maps are also not playable ?

Yes, all rest maps are crashes

IvanSavenko commented on lib/gameState/CGameState.cpp:
```diff
@@ -653,7 +670,16 @@ void CGameState::initHeroes(IGameRandomizer & gameRandomizer)
 	{
 		auto hero = getHero(heroID);
-		assert(map->isInTheMap(hero->visitablePos()));
+		const auto pos = hero->visitablePos();
+
+		if(!map->isInTheMap(pos))
+		{
+			logGlobal->warn(
+				"initHeroes: hero %s has invalid visitablePos %s (outside map) – skipping boat generation",
+				hero->getNameTranslated(), pos.toString());
+			continue;
+		}
+
```
How such heroes will be handled afterwards? I assume they can't even move or be interacted with from their position

MichalZr6 replied:
Fair point. All such misplaced visitable objects are now repositioned during map init.

IvanSavenko commented on lib/CPlayerState.h:
```diff
@@ -141,7 +143,14 @@ class DLL_LINKAGE PlayerState : public CBonusSystemNode, public Player, public G
+		if(h.hasFeature(Handler::Version::CONTROL_LOSS_TRACKING))
+			h & everControlledObjects;
+		else if(!h.saving)
+			everControlledObjects.clear();
```
FYI - this block is unnecessary. When deserializing (not loading) all objects are always freshly created, so this set is guaranteed to be empty

MichalZr6 replied:
Ok. Unnecessary so removed now.

IvanSavenko commented on lib/mapping/MapFormatH3M.cpp:
```diff
@@ -479,8 +479,8 @@ void CMapLoaderH3M::readVictoryLossConditions()
-				oper.expressions.emplace_back(EventCondition(EventCondition::CONTROL, 0, Obj(Obj::CREATURE_GENERATOR1)));
-				oper.expressions.emplace_back(EventCondition(EventCondition::CONTROL, 0, Obj(Obj::CREATURE_GENERATOR4)));
+					oper.expressions.emplace_back(EventCondition(EventCondition::CONTROL_CURRENT, 0, Obj(Obj::CREATURE_GENERATOR1)));
+					oper.expressions.emplace_back(EventCondition(EventCondition::CONTROL_CURRENT, 0, Obj(Obj::CREATURE_GENERATOR4)));
```
Looks like wrong indentation here in this method? (and similar blocks before & after)

MichalZr6 replied:
Fixed.

IvanSavenko reviewed (approved):
Minor issues (check review), but nothing critical.

Can merge as it, or wait for these minor changes

MichalZr6 commented:
Thanks for review. I will fix those in next few days.

MichalZr6 committed:
Adjust visitablePos for objects that are misplaced on map (out of map bounds)

MichalZr6 committed:
Warn in mapeditor's validator about objects that are placed out of map bounds

MichalZr6 commented:
I think I've fixed all the review points. Updated PR description and - after testing the most problematic maps - comments in map's table. PR should be ready now.

IvanSavenko commented:
Please  resolve conflicts and I'll merge this one

MichalZr6 commented:
> Please resolve conflicts and I'll merge this one

Done
